### PR TITLE
feat(gatsby): Add preliminary fast-refresh integration

### PR DESCRIPTION
**Pull request by blainekasten**
## Description

TL;DR: Custom overlay, removing page query export, adding some tests

---

Fast Refresh is supported with React 16.8+. Previously we've only been using FastRefresh if users specifically set `GATSBY_HOT_LOADER=fast-refresh`. The problem here is that React 17 does not work with `react-hot-loader` and so the better option here is to detect if the react version supports fast refresh.

~This PR is to sniff into react's version and detect if fast refresh is usable. The user is still able to override it with `GATSBY_HOT_LOADER=react-hot-loader`.~
We'll revisit this in another PR.

This PR also ensures that page queries are removed via babel because with FastRefresh, updates scan upwards through their module tree to scan for usages to know what modules all need updating. With page queries there is a big problem because they are never imported, the end effect is then that the FastRefresh can't know if it's safe to update and will reject the update requiring the user to reload. This is bad because page queries are quite common.

~This is potentially problematic with https://github.com/gatsbyjs/gatsby/pull/20672, so we might need to address that before this can be merged.~ [nvm](https://github.com/gatsbyjs/gatsby/pull/20672#issuecomment-690095014)

This PR/implementation is not perfect and most likely will be refactored when we release it for everyone, until then this is good enough. Some TODOs were left in the code on purpose, not needed to be addressed in this PR.

### Documentation

Only when Fast Refresh is available for everyone (as we e.g. bump the minimum React version) we need to change docs.

## Related Issues

[ch19063]

### CODE REVIEW

File: packages/gatsby/src/utils/detect-hot-loader-to-use.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  // If the user has defined this, we don't want to do any package sniffing
+  if (process.env.GATSBY_HOT_LOADER) return
+
+  // Do some package sniffing to see if we can use fast-refresh if the user didn't
+  // specify a specific hot loader with the environment variable.
+  try {
+    const reactVersion = require(`react/package.json`).version
+    const parsed = semver.parse(reactVersion)
+
+    console.log(parsed)
```
Comment (smthomas) [inline packages/gatsby/src/utils/detect-hot-loader-to-use.ts:16]: Not a lot of context on this but it looks good to me. However, should this and the console.debug below have been removed?

File: packages/gatsby/src/services/initialize.ts [hunk 2]
```diff
@@ -52,6 +53,8 @@ export async function initialize({
     reporter.panic(`Missing program args`)
   }
 
+  detectHotLoaderToUse()
```
Comment (wardpeet) [inline packages/gatsby/src/services/initialize.ts:56]: I would like these to be super long and meaningful cause it's not just detecting because I would assume a `const hotLoader = detectHotLoaderToUse()` instead
```suggestion
  detectAndSetWhichReactHotLoaderToUse()
```

File: packages/gatsby/src/utils/detect-hot-loader-to-use.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    const reactVersion = require(`react/package.json`).version
+    const parsed = semver.parse(reactVersion)
+
+    if (
+      parsed &&
+      // React 17+ is supported
+      (parsed.major >= 17 ||
+        // React 16.8+ is supported
+        (parsed.major === 16 && parsed.minor >= 8))
+    ) {
```
Comment (wardpeet) [inline packages/gatsby/src/utils/detect-hot-loader-to-use.ts:22]:
```suggestion
    if (semver.satisfies(reactVersion, '>=16.8.0')) {
```

File: e2e-tests/development-runtime/package.json [hunk 4]
```diff
@@ -40,7 +40,7 @@
     "update:preview": "curl -X POST -d '{ \"fake-data-update\": true }' -H \"Content-Type: application/json\" http://localhost:8000/__refresh",
     "start-server-and-test": "start-server-and-test develop http://localhost:8000 cy:run",
     "cy:open": "cypress open",
-    "cy:run": "(is-ci && cypress run --browser chrome --record) || cypress run --browser chrome"
+    "cy:run": "(is-ci && cypress run --browser chrome --record)"
```
Comment (blainekasten) [inline e2e-tests/development-runtime/package.json:43]: did you know that if e2e tests failed in CI that we ran them again because the exit code would be 1 and trigger the || statement 👎

Comment (pieh) [reply e2e-tests/development-runtime/package.json:43]: ooops, that's on me I think ... goal here was to use `--record` only in CI

Comment (pieh) [reply e2e-tests/development-runtime/package.json:43]: I think this just won't run locally? so it still need some love?

Comment (blainekasten) [reply e2e-tests/development-runtime/package.json:43]: Does anyone really need to run this locally? Usually I use `cy:open` locally

Comment (blainekasten) [reply e2e-tests/development-runtime/package.json:43]: so yeah i guess I sort of broke functionality, i can improve if you think it's worth it

Comment (pieh) [reply e2e-tests/development-runtime/package.json:43]: Can we just drop this change from this PR and handle this separately? It just feels like this doesn't belong in this PR

Comment (pieh) [reply e2e-tests/development-runtime/package.json:43]: I'll take care of doing this so it doesn't run twice on failures in CI + still runs fine locally

File: packages/gatsby-cli/package.json [hunk 5]
```diff
@@ -37,7 +37,7 @@
     "pretty-error": "^2.1.1",
     "progress": "^2.0.3",
     "prompts": "^2.3.2",
-    "react": "^16.8.0",
+    "react": "^16.12.0",
```
Comment (pieh) [inline packages/gatsby-cli/package.json:40]: Is this needed in this PR? https://github.com/gatsbyjs/gatsby/pull/21522 has some info why this is intentionally lowest it can be

File: packages/gatsby/src/utils/webpack.config.js [hunk 6]
```diff
@@ -164,11 +164,12 @@ module.exports = async (
         return {
           polyfill: directoryPath(`.cache/polyfill-entry`),
           commons: [
-            `${require.resolve(
-              `webpack-hot-middleware/client`
-            )}?path=${getHmrPath()}`,
+            process.env.GATSBY_HOT_LOADER !== `fast-refresh` &&
```
Comment (wardpeet) [inline packages/gatsby/src/utils/webpack.config.js:167]: Do we need hmr middleware?

File: packages/gatsby/src/utils/get-react-hot-loader-strategy.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  // specify a specific hot loader with the environment variable.
+  try {
+    const reactVersion = require(`react/package.json`).version
+
+    if (semver.satisfies(reactVersion, `>=16.9.0`)) {
+      return `fast-refresh`
+    }
+  } catch (e) {
+    return `react-hot-loader`
+  }
```
Comment (wardpeet) [inline packages/gatsby/src/utils/get-react-hot-loader-strategy.ts:20]:
```suggestion
  // try {
  //   const reactVersion = require(`react/package.json`).version

  //   if (semver.satisfies(reactVersion, `>=16.9.0`)) {
  //     return `fast-refresh`
  //   }
  // } catch (e) {
  //   return `react-hot-loader`
  // }
```

File: packages/gatsby/src/utils/get-react-hot-loader-strategy.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+export function getReactHotLoaderStrategy(): string {
+  // If the user has defined this, we don't want to do any package sniffing
+  if (process.env.GATSBY_HOT_LOADER) return process.env.GATSBY_HOT_LOADER
+
+  // Do some package sniffing to see if we can use fast-refresh if the user didn't
+  // specify a specific hot loader with the environment variable.
+  try {
+    const reactVersion = require(`react/package.json`).version
+
+    if (semver.satisfies(reactVersion, `>=16.9.0`)) {
```
Comment (wardpeet) [inline packages/gatsby/src/utils/get-react-hot-loader-strategy.ts:15]:
```suggestion
    // TODO React components need to be named to make fast-refresh work
    // We need to create an eslint-rule that shows this error or create a babel plugin
    // that converts arrow functions to generated named ones.
    // When it's available we can switch the condition to >=16.9.0
    if (semver.satisfies(reactVersion, `>=17.0.0`)) {
```

**Comment by blainekasten** [comment]
Not sure how to move forward with this now. The tests won't pass while multiple versions of React are installed (e.g., installing React 16.9+ for the e2e tests to use fast refresh while gatsby-cli installs 16.8). We need to find a way to decouple the react version from gatsby-cli. I'm afraid we're only going to run into more and more of these problems as we work to support React 17. 

is this now proving to be a requirement for Gatsby v3?

**Comment by pieh** [comment]
I just opened https://github.com/gatsbyjs/gatsby/pull/26887 that removes `react` from `gatsby-cli` dependencies (moves to devDeps and bundle it) which should let us move past those problems.

**Comment by blainekasten** [comment]
@pieh awesome! Ping me when it goes through and I'll update this branch and hopefully the tests pass then!

**Review by pieh — APPROVED** [review]
Yes! 🎉 

We still have to make decision on enabling by default for react@17, but we shouldn't block all the improvements done here on that.
